Assets/Scripts/Layers/Adapters/ChainOfResponsibility.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using UnityEngine;
+using UnityEngine.Events;
+using UnityEngine.Networking;
+
+namespace Netherlands3D.Twin
+{
+    public class ChainOfResponsibility : MonoBehaviour
+    {
+        [Header("Data type adapters")] [Space(5)]
+        [SerializeField] private ScriptableObject[] dataTypeAdapters;
+        private IDataTypeAdapter[] dataTypeAdapterInterfaces;
+        public UnityEvent OnAdapterFound = new();
+
+        [Header("Events invoked on failures")] [Space(5)]
+        public UnityEvent<string> CouldNotFindAdapter = new();
+        public UnityEvent<string> OnDownloadFailed = new();
+
+        private string targetUrl = "";
+        public string TargetUrl { get => targetUrl; set => targetUrl = value; }
+
+        private Coroutine chain;
+
+        private void OnDisable() {
+            if(chain != null)
+                StopCoroutine(chain);
+        }
+
+        /// <summary>
+        /// Determine the type of data using chain of responsibility
+        /// </summary>
+        /// <param name="url">Url to file or service</param>
+        public void DetermineAdapter(string url)
+        {
+            TargetUrl = url;
+            DetermineAdapter();
+        }
+        public void DetermineAdapter()
+        {
+            if(chain != null)
+                StopCoroutine(chain);
+
+            chain = StartCoroutine(DownloadAndCheckSupport(TargetUrl));
+        }
+
+        private IEnumerator DownloadAndCheckSupport(string url)
+        {
+            // Start by download the file, so we can do a detailed check of the content to determine the type
+            var urlAndData = new LocalFile()
+            {
+                SourceUrl = url,
+                LocalFilePath = ""
+            };
+            yield return DownloadDataToLocalCache(urlAndData);
+
+            // No local cache? Download failed.
+            if(string.IsNullOrEmpty(urlAndData.LocalFilePath))
+            {
+                OnDownloadFailed.Invoke(url);
+                yield break;
+            }
+
+            // Find the proper adapter in a chain of responsibility
+            yield return AdapterChain(urlAndData);
+        }
+
+        /// <summary>
+        /// We download the file to a local cache so we can check the content from the adapters using streamreading.
+        /// This way we keep the heap memory usage low and can handle large files (like large obj's, or large WFS responses)
+        /// </summary>
+        /// <param name="urlAndData">The local file object where the path will set</param>
+        /// <returns></returns>
+        private IEnumerator DownloadDataToLocalCache(LocalFile urlAndData)
```

Comment:
Would it make sense to move this into the LocalFile class? That if you attempt to get the contents that it will do this on demand?

Reply:
Hmm we could do that, but then you'd have to wait for some sort of callback.
Now all adapters can do their Support checks without having to wait for any async stuff.

Reply:
Same as https://github.com/Netherlands3D/twin/pull/198/files/0d1825fb64c829166cc3c8cdafd316b41a8989f1#diff-05722b960df50c1798a23802bf2fb02e5f7e92d6fa1247d2c885f8fd0921bdc8